### PR TITLE
Advanced usage docs

### PR DESCRIPTION
**Pull request by zacespinosa**
Fixed some bugs, added advanced usage docs, new handling for lost qubits, simulation returns list of programs, removed network_monitor, noise.measure creates one ro

Let me know if any of the explanations are confusing and if the code examples can be more clear or cleaned up. Hit me with that GRE essay writing

### CODE REVIEW

File: docs/demos/advanced-usage.rst [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+In classical networking this would include devices such as a router, modem, repeater, 
+and network switch. In quantum networking think fiber optics, intensity modulators, lasers or 
+photon sources, and `SNSPDs <https://en.wikipedia.org/wiki/Superconducting_nanowire_single-photon_detector>`_ or detectors. 
+Associated with each device in a quantum network is its physical effect on qubits and the time it takes 
+qubits to pass through the device.
+
+There are three types of devices in netQuil: source, transit, and target devices. Source and target devices
+are associated with an agent, while transit devices are associated with instances of``QConnection``. When a qubit
+leaves Alice and travels to Bob, the qubit originates from Alice's source devices, travels through the
+transit devices attached to their ``QConnection`` and ends by hitting Bob's target devices. The
```
Comment (mattradz) [inline docs/demos/advanced-usage.rst]: "arriving through" maybe instead of hitting

File: docs/demos/advanced-usage.rst [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+Built-in Devices
+================
+``Laser`` is a built-in noisy photon source that generates photons according to ``expected_photons`` and a poisson 
+distribution. If ``network_monitor=True`` the laser will output its noise to signal ratio after each trial. You can create a custom ``get_results`` function that prints
+information about the device after each trial. The ``Laser`` will also return a delay equal to the photon
+pulse length. 
+
+``Fiber`` is a built-in noisy fiber optical wire simulator and an example of a transit device. Fibers
+have an associated length in kilometers and an attenutation coefficient. The attenuation coefficient is
```
Comment (mattradz) [inline docs/demos/advanced-usage.rst]: The attenuation coefficient is "proportional to", i think its prob of loss = 10^(( length*attenuation coefficient)/10)

File: docs/demos/advanced-usage.rst [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+``Laser`` is a built-in noisy photon source that generates photons according to ``expected_photons`` and a poisson 
+distribution. If ``network_monitor=True`` the laser will output its noise to signal ratio after each trial. You can create a custom ``get_results`` function that prints
+information about the device after each trial. The ``Laser`` will also return a delay equal to the photon
+pulse length. 
+
+``Fiber`` is a built-in noisy fiber optical wire simulator and an example of a transit device. Fibers
+have an associated length in kilometers and an attenutation coefficient. The attenuation coefficient is 
+equal to the probability that a photon is lost while traveling within the fiber (i.e. the photon is measured, 
+and the measured value is inaccessible by any agent). In netQuil, if a qubit is lost due to attenuation, the 
+target will receive the negative of the qubit lost. For example, if Alice sends qubit 3 and it is
```
Comment (mattradz) [inline docs/demos/advanced-usage.rst]: "the negative index of the qubit lost" rather than the "index". Maybe at text about how we are sending indices or this may be known from our description earlier.

File: docs/demos/advanced-usage.rst [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    QConnect(alice, bob, transit_devices=[fiber])
+
+    Simulation(alice, bob).run(network_monitor=True)
+
+    qvm = QVMConnection()
+    results = qvm.run(p)
+    print(results)
+
+NetQuil also has a built-in noise module for performing common qubit operations such as normal 
+unitary rotations, depolarization, and phase flips.
```
Comment (mattradz) [inline docs/demos/advanced-usage.rst]: and "bit and phase flips"

File: docs/demos/advanced-usage.rst [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+Trials and Time
+===============
+In some situation, pyQuil programs generated between trials will be different depending
```
Comment (mattradz) [inline docs/demos/advanced-usage.rst]: situations

File: netQuil/devices.py [hunk 6]
```diff
@@ -29,21 +31,27 @@ def apply(self, program, qubits):
...
+        Prints device information about trial to console. Should use tqdm.write in order to 
+        avoid conflicts with network monitor. This function will only run when verbose 
+        for the device is set to True.
+        '''
+        pass
     
     def reset(self): 
-        self.success = 0
-        self.trials = 0 
+        '''
```
Comment (mattradz) [inline netQuil/devices.py]: Does this just give them the outline of methods they can implement?

Comment (zacespinosa) [reply netQuil/devices.py]: yeah, same at get_results and apply. They're just functions that all child classes of device have.

File: demos/advancedUsage.py [hunk 7]
```diff
@@ -93,8 +22,7 @@ def __init__(self, length, fiber_quality, rotation_std):
 
     def apply(self, program, qubits):
         for qubit in qubits: 
-            if qubit < 0: continue
```
Comment (mattradz) [inline demos/advancedUsage.py]: Why do we not need this check?

File: demos/advancedUsage.py [hunk 8]
```diff
@@ -112,13 +39,14 @@ def run(self):
         for q in self.qubits:
             p += H(q)
             p += X(q)
-            self.qsend('Bob', [q])
+            self. qsend('Bob', [q])
```
Comment (mattradz) [inline demos/advancedUsage.py]: Remove the space?

File: docs/demos/advanced-usage.rst [hunk 9]
```diff
@@ -88,9 +92,9 @@ Custom Devices
 ==============
 Moreover, you can build your own custom devices by extending the ``Device`` class. All devices must have
 an ``apply`` function that is responsible for the device's activity. The ``run`` function must return a 
-dictionary that optionally contains the qubits and delay. The delay represents the time it took qubits to 
-travel through the device. Remember to switch the sign of a qubit if it is lost during transmission, as well as 
-add the list of qubits to the dictionary that is returned.
+dictionary that optionally contains the lost qubits and delay. The delay represents the time it took qubits to 
+travel through the device. Remember, if qubits are lost while passing through the device, return an entry in the dictionary 
+``lost_qubits: [lost qubits]``.
```
Comment (mattradz) [inline docs/demos/advanced-usage.rst]: i like this

File: demos/advancedUsage.py (hunk 7 above)
Comment (zacespinosa) [reply demos/advancedUsage.py]: Because we are no longer passing lost qubits to each device, so they don't need to do that check. We pass qubits to agent after going through all devices, just not to every device.
